### PR TITLE
External-dns and new clusterissuer dns01 Cloudflare

### PR DESCRIPTION
**Pull request by mrkhachaturov**
Overview

This pull request introduces the integration of External-DNS into the full bundles and adds support for a dns01 ClusterIssuer using Cloudflare. It enhances the DNS management capabilities for our deployments by allowing dynamic DNS record management directly from Kubernetes resources.

Changes Made

 1. **External-DNS Integration:**
 
   - Added External-DNS to the full deployment bundles.

  - Configured External-DNS to automatically manage DNS records for services within the Kubernetes cluster ( we must discuss how to configure external-dns via configmap or create an application in tenant `external-dns` where we can define values).

We must define some additional annotations for ingresses in order to make external-dns work , so we must discuss this also which is best method to configure it ( from configmap or dashboard ).

**2. dns01 ClusterIssuer for Cloudflare:**

- Implemented support for a dns01 ClusterIssuer using Cloudflare.
- This allows for automated certificate issuance via DNS challenge, leveraging Cloudflare as the DNS provider.
- The configuration can be defined in the Cozystack ConfigMap

3. Default Ingress Configuration: 

 - Updated the default Ingress resources to use Cloudflare for DNS challenges.
- Ensured that if the Cloudflare issuer is defined in the Cozystack ConfigMap, it will be utilized for all default Ingresses, streamlining the deployment process and improving reliability.

**Benefits**

- Automated DNS Management: With External-DNS, DNS entries will be created and updated automatically based on the state of Kubernetes resources, reducing manual overhead.
- Seamless Certificate Management: The dns01 ClusterIssuer integration allows for automated SSL/TLS certificate issuance, enhancing security for deployed applications.
- Flexibility in Configuration: Users can easily switch between different issuers by updating the Cozystack ConfigMap, providing flexibility in the choice of DNS and certificate management solutions.

## Release Notes

- **New Features**
  - Introduced a new `external-dns` release with support for managing DNS records in Kubernetes.
  - Added configuration options for DNS synchronization policies and provider settings.
  - Implemented a new lookup for issuer types in Ingress configurations.
  - Expanded configuration with new entries for `external-dns` in multiple deployment files, enhancing deployment flexibility.

- **Documentation**
  - Comprehensive README and configuration schema for the `external-dns` Helm chart added, detailing installation and customization options.

- **Improvements**
  - Enhanced RBAC configuration for flexible permissions management.
  - Updated annotations and health check configurations for better service monitoring.

### CODE REVIEW

File: packages/extra/ingress/templates/dashboard.yaml [hunk 1]
```diff
@@ -1,29 +1,36 @@
...
+
+{{- if .Values.dashboard }}  
+apiVersion: networking.k8s.io/v1  
+kind: Ingress  
+metadata:  
+  annotations:  
+    cert-manager.io/cluster-issuer: letsencrypt-prod  
+    {{- if eq $issuerType "cloudflare" }} 
+    {{- else }}  
+    acme.cert-manager.io/http01-ingress-class: tenant-root
```
Comment (kvaps) [inline packages/extra/ingress/templates/dashboard.yaml]: It seems we have a bug here. Could you please update to 
```suggestion
    acme.cert-manager.io/http01-ingress-class: {{ .Release.Namespace }}  
```

File: packages/core/platform/bundles/distro-full.yaml [hunk 2]
```diff
@@ -140,3 +140,9 @@ releases:
   chart: cozy-telepresence
   namespace: cozy-telepresence
   dependsOn: []
+
+- name: external-dns
+  releaseName: external-dns
+  chart: cozy-external-dns
+  namespace: cozy-external-dns
+  dependsOn: [cilium,cert-manager]
```
Comment (kvaps) [inline packages/core/platform/bundles/distro-full.yaml]: I think we also need to add it to `distro-full` and `distro-hosted` bundles.

**Review by kvaps — COMMENTED** [review]
Thanks for your contribution! Just few momets I mentioned here:

File: packages/core/platform/bundles/distro-full.yaml (hunk 2 above)
Comment (mrkhachaturov) [reply packages/core/platform/bundles/distro-full.yaml]: Sure, added it to distro-hosted as well

**Review by kvaps — APPROVED** [review]
LGTM, thank you

File: packages/system/cert-manager-issuers/templates/cluster-issuers.yaml [hunk 3]
```diff
@@ -1,35 +1,56 @@
...
----
-apiVersion: cert-manager.io/v1
-kind: ClusterIssuer
-metadata:
-  name: letsencrypt-stage
-spec:
-  acme:
-    privateKeySecretRef:
+{{- $cozyConfig := lookup "v1" "ConfigMap" "cozy-system" "cozystack" }}  
+{{- $issuerType := (index $cozyConfig.data "clusterissuer") | default "http01" }}
```
Comment (kvaps) [inline packages/system/cert-manager-issuers/templates/cluster-issuers.yaml:2]: @mrkhachaturov could you please update documentation on how to use this parameter

https://github.com/aenix-io/cozystack-website/

**Pull request merged**